### PR TITLE
Move device preprocessing to inner execute

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**
If there is a gradient transform, we want device preprocessing to happen within the ml boundary, i.e., in `inner_execute`, to avoid applying the gradient transform on transformed tapes produced by device preprocessing.

**Description of the Change:**
1. Adds a `inner_transform` keyword to `qml.execute`
2. Update `QNode._execute_component` and `qml.execute` to add device preprocessing to the inner transform program if the `gradient_fn` is a gradient transform.

**Related GitHub Issues:**
[sc-59107]

### CODE REVIEW

File: pennylane/devices/preprocess.py [hunk 1]
```diff
@@ -162,7 +165,7 @@ def mid_circuit_measurements(
         mcm_method = "one-shot" if tape.shots else "deferred"
 
     if mcm_method == "one-shot":
-        return qml.dynamic_one_shot(tape)
+        return qml.dynamic_one_shot(tape, interface=interface)
```
Comment (astralcai) [inline pennylane/devices/preprocess.py:168]: The `dynamic_one_shot` transform is created outside the ml boundary but applied in inner execute, after `convert_to_numpy_parameters`, which erases the original interface information. This allows the transform to remember what the interface was, and this is important for hardware like post selection.

File: pennylane/ops/op_math/condition.py [hunk 2]
```diff
@@ -28,14 +28,14 @@ class ConditionalTransformError(ValueError):
     """Error for using qml.cond incorrectly"""
 
 
-class Conditional(SymbolicOp):
+class Conditional(SymbolicOp, Operation):
```
Comment (astralcai) [inline pennylane/ops/op_math/condition.py:31]: Making `Conditional` also a subclass of `Operation` such that it inherits `grad_method`, which is required for the gradient transform.

File: pennylane/transforms/dynamic_one_shot.py [hunk 3]
```diff
@@ -228,7 +233,7 @@ def measurement_with_no_shots(measurement):
             else np.nan
         )
 
-    interface = qml.math.get_deep_interface(circuit.data)
+    interface = interface or qml.math.get_deep_interface(circuit.data)
```
Comment (astralcai) [inline pennylane/transforms/dynamic_one_shot.py:236]: Inside inner execution, `circuit.data` has already been converted to numpy. The original interface was passed into the transform via the added keyword argument and used here.

File: pennylane/transforms/core/transform_program.py [hunk 4]
```diff
@@ -354,24 +352,32 @@ def set_classical_component(self, qnode, args, kwargs):
...
-        trans_type = np.zeros(len(self._transform_program), dtype=np.int32)
-        for i, t in enumerate(self._transform_program):
-            if "dynamic_one_shot" in str(t):
-                trans_type[i] = 1
-            if "mid_circuit_measurements" in str(t):
-                trans_type[i] = 2
-        if sum(trans_type) < 2:
-            return
-        keep = 2 if 2 in trans_type else 1
+    def _prune_dynamic_transform(self, type_to_keep=1):
```
Comment (albi3ro) [inline pennylane/transforms/core/transform_program.py:355]: Given this is a public method on a key architectual component, I don't think we should rename/ remove it without a deprecation cycle.  @vincentmr any thoughts?

Comment (astralcai) [reply pennylane/transforms/core/transform_program.py:355]: I figured that it wasn't public on purpose, since it is only used internally in `qml.execute` to make sure that we don't end up having duplicate `dynamic_one_shot` transforms. The only time that this would happen is when the user manually applies `dynamic_one_shot` on a circuit, and device preprocessing adds another `dynamic_one_shot`, which is only internal.

File: pennylane/transforms/core/transform_program.py [hunk 5]
```diff
@@ -545,3 +551,23 @@ def __call__(self, tapes: Tuple[QuantumTape]) -> Tuple[ResultBatch, BatchPostPro
...
+    elif any("dynamic_one_shot" in str(t) for t in all_transforms):
+        type_to_keep = 1
+
+    if type_to_keep == 0:
+        return
+
+    dynamic_transform_found = inner_transform._prune_dynamic_transform(type_to_keep)
+    if dynamic_transform_found:
+        type_to_keep = 0
+    outer_transform._prune_dynamic_transform(type_to_keep)
```
Comment (dwierichs) [inline pennylane/transforms/core/transform_program.py:573]: Could you add some code comments regarding the superficial behaviour between `mid_circuit_measurements` and `dynamic_one_shot` and between inner and outer transform?

Nice implementation of the logic!

File: pennylane/workflow/execution.py [hunk 6]
```diff
@@ -253,7 +253,13 @@ def device_expansion_function(tape):  # pylint: disable=function-redefined
...
+    device,
+    override_shots,
+    cache,
+    inner_transform,
+    expand_fn=None,
+    execution_config=None,
+    numpy_only=True,
 ) -> Callable:
     """Construct the function that will execute the tapes inside the ml framework registration
     for the 1st order derivatives.
```
Comment (dwierichs) [inline pennylane/workflow/execution.py:265]: Is the docstring up to date with this change? :thinking:

File: pennylane/workflow/execution.py [hunk 7]
```diff
@@ -435,6 +443,7 @@ def execute(
             This affects the types of parameters that can exist on the input tapes.
             Available options include ``autograd``, ``torch``, ``tf``, ``jax`` and ``auto``.
         transform_program(.TransformProgram): A transform program to be applied to the initial tape.
+        inner_transform (.TransformProgram): A transform program to be applied to the tapes in the inner_execute.
```
Comment (dwierichs) [inline pennylane/workflow/execution.py:446]: Might be good to mention the relationship to the ML interface?

File: tests/transforms/test_experimental/test_transform_program.py [hunk 8]
```diff
@@ -107,6 +108,27 @@ def postprocessing2(results):
...
+                qml.transforms.dynamic_one_shot,
+            ]
+        )
+        program2 = TransformProgram(
+            [
+                qml.transforms.dynamic_one_shot,
+                qml.transforms.sum_expand,
+            ]
+        )
+
```
Comment (dwierichs) [inline tests/transforms/test_experimental/test_transform_program.py:127]: Could include a test for the priority between `mid_circuit_measurements` and `dynamic_one_shot`? :thinking:

**Review by dwierichs — CHANGES_REQUESTED** [review]
This looks great, thanks @astralcai ! Only had a few tiny comments/questions :)

File: pennylane/workflow/qnode.py [hunk 9]
```diff
@@ -28,6 +28,7 @@
 from pennylane.logging import debug_logger
 from pennylane.measurements import CountsMP, MidMeasureMP, Shots
 from pennylane.tape import QuantumScript, QuantumTape
+from pennylane.transforms.core.transform_program import _prune_dynamic_transform
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:31]: I'd prefer not to have a coupling between the transform program and a high level workflow class via a private method.  I know `prune_dynamic_transform` is kinda a patch, but I do think it needs to be a public patch.

Comment (astralcai) [reply pennylane/workflow/qnode.py:31]: How about I move this function into `_qnode.py`? The transform program shouldn't need to know what inner transform and outer transform are anyways?

File: tests/transforms/test_experimental/test_transform_program.py (hunk 8 above)
Comment (astralcai) [reply tests/transforms/test_experimental/test_transform_program.py]: Done

File: pennylane/transforms/core/transform_program.py (hunk 4 above)
Comment (albi3ro) [reply pennylane/transforms/core/transform_program.py:355]: So there's public as in "user facing" , and public as in "accessible outside the class".

Comment (astralcai) [reply pennylane/transforms/core/transform_program.py:355]: True, I can make this public again.

File: pennylane/workflow/qnode.py (hunk 9 above)
Comment (albi3ro) [reply pennylane/workflow/qnode.py]: I actually like that idea 👍 This does seem to be more about "how do I setup and validate a workflow" than maintaining and executing a generic queue of transforms.

File: pennylane/devices/default_qubit.py [hunk 10]
```diff
@@ -502,7 +502,10 @@ def preprocess(
 
         transform_program.add_transform(validate_device_wires, self.wires, name=self.name)
         transform_program.add_transform(
-            mid_circuit_measurements, device=self, mcm_config=config.mcm_config
+            mid_circuit_measurements,
+            device=self,
+            mcm_config=config.mcm_config,
+            interface=config.interface,
```
Comment (mudit2812) [inline pennylane/devices/default_qubit.py:508]: Note that the same change needs to be made to `lightning.qubit` and `lightning.kokkos` preprocessing. We should open a PR in lightning in parallel to avoid errors once this PR is merged.

File: pennylane/ops/op_math/condition.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/ops/op_math/condition.py:31]: We should probably double check the status of parameter shift + conditionals.  If we had previously been supporting parameter shift + conditionals due to defer measurements being applied earlier, we may need some to think about what we want to do for it.

File: pennylane/workflow/qnode.py [hunk 11]
```diff
@@ -1051,29 +1051,35 @@ def _execution_component(self, args: tuple, kwargs: dict, override_shots) -> qml
...
             and hasattr(self.device, "capabilities")
             and self.device.capabilities().get("supports_mid_measure", False)
         )
         if has_mcm_support:
-            full_transform_program.add_transform(
+            inner_transform_program.add_transform(
                 qml.devices.preprocess.mid_circuit_measurements,
                 device=self.device,
                 mcm_config=self.execute_kwargs["mcm_config"],
+                interface=self.interface,
```
Comment (mudit2812) [inline pennylane/workflow/qnode.py:1078]: This should cover `lightning.kokkos`'s mid-circuit measurement pipeline 😄

Comment (astralcai) [reply pennylane/workflow/qnode.py:1078]: Is `lightning.kokkos` a legacy device? If so then yes I guess.

File: pennylane/transforms/core/transform_program.py [hunk 12]
```diff
@@ -354,24 +352,32 @@ def set_classical_component(self, qnode, args, kwargs):
...
-        for i, t in enumerate(self._transform_program):
-            if "dynamic_one_shot" in str(t):
-                trans_type[i] = 1
-            if "mid_circuit_measurements" in str(t):
-                trans_type[i] = 2
-        if sum(trans_type) < 2:
-            return
-        keep = 2 if 2 in trans_type else 1
+    def prune_dynamic_transform(self, type_to_keep=1):
+        """Ensures that only one or none ``dynamic_one_shot`` is applied.
```
Comment (trbromley) [inline pennylane/transforms/core/transform_program.py:356]: The name and first line don't seem to match what the method does. Would `check_dynamic_transform` be more appropriate?

Comment (astralcai) [reply pennylane/transforms/core/transform_program.py:356]: This method does prune the transform program, removing any duplicate `dynamic_one_shot` transforms. This is an existing public method, so I didn't want to change its name.

Comment (albi3ro) [reply pennylane/transforms/core/transform_program.py:356]: and this name was set last release cycle. So I'm not sure we can change it at this point.

File: pennylane/devices/default_qubit.py (hunk 10 above)
Comment (dwierichs) [reply pennylane/devices/default_qubit.py:508]: Has a PR been opened? Or is this tracked elsewhere?

File: pennylane/ops/op_math/condition.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/ops/op_math/condition.py:31]: What is the status here?

File: pennylane/transforms/core/transform_program.py (hunk 12 above)
Comment (dwierichs) [reply pennylane/transforms/core/transform_program.py:356]: >The name and first line don't seem to match what the method does.

Why is this? I seem to be overlooking the mismatch :thinking:

File: tests/test_qnode.py [hunk 13]
```diff
@@ -2011,3 +2012,47 @@ def circuit(x):
...
+    )
+    program2 = qml.transforms.core.TransformProgram(
+        [
+            qml.transforms.dynamic_one_shot,
+            qml.transforms.sum_expand,
+        ]
+    )
+
+    _prune_dynamic_transform(program1, program2)
+    assert len(program1) == 2
```
Comment (dwierichs) [inline tests/test_qnode.py:2057]:
```suggestion
    assert len(program1) == 2
    assert qml.transforms.mid_circuit_measurements in program1
```

**Review by dwierichs — COMMENTED** [review]
Looks good to me, thanks @astralcai !
Just had a small suggestion for one test, and wanted to follow up on open comments.

File: pennylane/ops/op_math/condition.py (hunk 2 above)
Comment (astralcai) [reply pennylane/ops/op_math/condition.py:31]: Right now existing support is maintained.

File: pennylane/devices/default_qubit.py (hunk 10 above)
Comment (astralcai) [reply pennylane/devices/default_qubit.py:508]: I believe there is a story on shortcut for it, but not a PR yet? @mudit2812

**Review by dwierichs — APPROVED** [review]
Thanks @astralcai

File: pennylane/workflow/qnode.py [hunk 14]
```diff
@@ -1157,3 +1165,31 @@ def __call__(self, *args, **kwargs) -> qml.typing.Result:
...
+
+    Sometimes device preprocess contains a ``mid_circuit_measurements`` transform, which will
+    be added to the inner transform program. If the user then applies a ``dynamic_one_shot``
+    manually, it will duplicate the ``mid_circuit_measurements`` transform. This function ensures
+    that there is only one ``dynamic_one_shot`` transform in the outer and inner transform
+    programs combined.
+
+    """
+
+    all_transforms = outer_transform._transform_program + inner_transform._transform_program
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:1182]:
```suggestion
    all_transforms = outer_transform + inner_transform
```

File: tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py [hunk 15]
```diff
@@ -1374,6 +1376,7 @@ def test_state(self, dev, diff_method, grad_on_execution, device_vjp, interface,
         x = jax.numpy.array(0.543)
         y = jax.numpy.array(-0.654)
         if not dev.wires:
+            dev = copy.copy(dev)
             dev._wires = qml.wires.Wires([0, 1])  # pylint:disable=protected-access
```
[the remark this answers is not in the document]
Comment (astralcai) [reply tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:1380]: `dev` is used globally, and `dev._wires = qml.wires.Wires([0, 1])` alters the global state, which causes other tests to fail.

[the remark this answers is not in the document]
Comment (astralcai) [reply tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:1380]: If you meant removing the `if` block completely, then it would cause this test to fail.

File: tests/transforms/core/test_transform_program.py [hunk 16]
```diff
@@ -356,6 +356,13 @@ def test_empty_program(self):
         ):
             program.get_last()
 
+    def test_get_last(self):
```
Comment (albi3ro) [inline tests/transforms/core/test_transform_program.py:359]: Was this just... not tested?

Comment (astralcai) [reply tests/transforms/core/test_transform_program.py:359]: Yeah but for some reason codecov never complained before and complains now.

File: tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply tests/interfaces/default_qubit_2_integration/test_jax_jit_qnode_default_qubit_2.py:1380]: Let's add the copy for now, but it shouldn't cause the test to fail to not specify wires.  May be related to #5813 that I'm currently trying to fix.

File: tests/transforms/core/test_transform_program.py (hunk 16 above)
Comment (albi3ro) [reply tests/transforms/core/test_transform_program.py:359]: 😢

**Review by albi3ro — APPROVED** [review]
🎉

File: pennylane/devices/default_qubit.py (hunk 10 above)
Comment (mudit2812) [reply pennylane/devices/default_qubit.py:508]: There is a [story](https://app.shortcut.com/xanaduai/iteration/61839?group_by=owner_ids&ct_workflow=all) and [open PR](https://github.com/PennyLaneAI/pennylane-lightning/pull/760).
